### PR TITLE
First pass at #555. Lazily reading request body.

### PR DESCRIPTION
**Pull request by johnrengelman**
[Review on Reviewable](https://reviewable.io/reviews/ratpack/ratpack/720)

### CODE REVIEW

File: ratpack-pac4j/src/main/java/ratpack/pac4j/RatpackPac4j.java [hunk 1]
```diff
@@ -392,12 +391,13 @@ public static Operation logout(Context ctx) {
...
-    return ctx.get(Session.class).getData().map(session -> webContext(ctx, session));
+  public static Promise<RatpackWebContext> webContext(Context ctx) {
+    return ctx.get(Session.class).getData().flatMap(session -> webContext(ctx, session));
   }
 
-  private static RatpackWebContext webContext(Context ctx, SessionData sessionData) {
-    return new RatpackWebContext(ctx, sessionData);
+  private static Promise<RatpackWebContext> webContext(Context ctx, SessionData sessionData) {
+    return ExecControl.execControl().blocking(() -> new RatpackWebContext(ctx, ctx.getRequest().getBody().block(),
+      ctx.parse(Form.class).onError(t -> Execution.execution().promiseOf((Form) null)).block(), sessionData));
```
Comment (johnrengelman) [inline ratpack-pac4j/src/main/java/ratpack/pac4j/RatpackPac4j.java]: This is how we're meshing the new Promise API with Pac4j's API.

**Comment by johnrengelman** [comment]
This only addresses Step #1 from https://github.com/ratpack/ratpack/issues/555.

**Comment by johnrengelman** [comment]
Reviewed 31 of 31 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/ratpack/ratpack/720)

**Comment by alkemist** [comment]
Review status: all files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

<sup>**[ratpack-core/src/main/java/ratpack/form/Form.java, line 46 \[r1\]](https://reviewable.io:443/reviews/ratpack/ratpack/720#-Ju9a-VthLRKl3DHRSOd)** ([raw file](https://github.com/ratpack/ratpack/blob/eed95a008fd7532c3d9f9a9b380ebc6c550634b3/ratpack-core/src/main/java/ratpack/form/Form.java#L46)):</sup>
Let's add {@code} in here to remove the need to use HTML entities.

---

<sup>**[ratpack-core/src/main/java/ratpack/form/internal/FormDecoder.java, line 48 \[r1\]](https://reviewable.io:443/reviews/ratpack/ratpack/720#-Ju9bP7Z4PZq7msLtECQ)** ([raw file](https://github.com/ratpack/ratpack/blob/eed95a008fd7532c3d9f9a9b380ebc6c550634b3/ratpack-core/src/main/java/ratpack/form/internal/FormDecoder.java#L48)):</sup>
Shouldn't be using promises here. This should receive a TypedData and return a Form.

---

<sup>**[ratpack-core/src/main/java/ratpack/handling/internal/DefaultContext.java, line 322 \[r1\]](https://reviewable.io:443/reviews/ratpack/ratpack/720#-Ju9cNj9nJl-MW14KVrA)** ([raw file](https://github.com/ratpack/ratpack/blob/eed95a008fd7532c3d9f9a9b380ebc6c550634b3/ratpack-core/src/main/java/ratpack/handling/internal/DefaultContext.java#L322)):</sup>
I think we should just throw this. It's always hard to know when to make this call, but this is really not a recoverable situation, so throwing makes more sense to me.

---

<sup>**[ratpack-core/src/main/java/ratpack/http/client/HttpClient.java, line 76 \[r1\]](https://reviewable.io:443/reviews/ratpack/ratpack/720#-Ju9cgsSSFIJl84Rk3jL)** ([raw file](https://github.com/ratpack/ratpack/blob/eed95a008fd7532c3d9f9a9b380ebc6c550634b3/ratpack-core/src/main/java/ratpack/http/client/HttpClient.java#L76)):</sup>
Should render the promise instead.

---

<sup>**[ratpack-core/src/main/java/ratpack/http/internal/ContentTypeHandler.java, line 33 \[r1\]](https://reviewable.io:443/reviews/ratpack/ratpack/720#-Ju9crLZ4GQTVrRonnuT)** ([raw file](https://github.com/ratpack/ratpack/blob/eed95a008fd7532c3d9f9a9b380ebc6c550634b3/ratpack-core/src/main/java/ratpack/http/internal/ContentTypeHandler.java#L33)):</sup>
We don't need to read the body to get the content type. It's in the headers.

---

<sup>**[ratpack-core/src/main/java/ratpack/jackson/Jackson.java, line 192 \[r1\]](https://reviewable.io:443/reviews/ratpack/ratpack/720#-Ju9dM1PhOLFY2xHoEn8)** ([raw file](https://github.com/ratpack/ratpack/blob/eed95a008fd7532c3d9f9a9b380ebc6c550634b3/ratpack-core/src/main/java/ratpack/jackson/Jackson.java#L192)):</sup>
I wonder if we should update these to render the promise, instead of using .then()

---

<sup>**[ratpack-core/src/main/java/ratpack/parse/Parser.java, line 68 \[r1\]](https://reviewable.io:443/reviews/ratpack/ratpack/720#-Ju9bhQtZv0fwQbHV6_M)** ([raw file](https://github.com/ratpack/ratpack/blob/eed95a008fd7532c3d9f9a9b380ebc6c550634b3/ratpack-core/src/main/java/ratpack/parse/Parser.java#L68)):</sup>
Shouldn't be working with promises here, it makes for more awkward user API.

---

<sup>**[ratpack-manual/src/content/chapters/33-jackson.md, line 101 \[r1\]](https://reviewable.io:443/reviews/ratpack/ratpack/720#-Ju9e_IfVCAA0cYvNxX7)** ([raw file](https://github.com/ratpack/ratpack/blob/eed95a008fd7532c3d9f9a9b380ebc6c550634b3/ratpack-manual/src/content/chapters/33-jackson.md#L101)):</sup>
I think we should update these examples to render the promise.

---

<sup>**[ratpack-pac4j/src/main/java/ratpack/pac4j/RatpackPac4j.java, line 394 \[r1\]](https://reviewable.io:443/reviews/ratpack/ratpack/720#-Ju9aJUIeb-cdxFx5fbH)** ([raw file](https://github.com/ratpack/ratpack/blob/eed95a008fd7532c3d9f9a9b380ebc6c550634b3/ratpack-pac4j/src/main/java/ratpack/pac4j/RatpackPac4j.java#L394)):</sup>
This is an internal type, it can't be used in public signatures.

---

<sup>**[ratpack-pac4j/src/main/java/ratpack/pac4j/RatpackPac4j.java, line 400 \[r1\]](https://reviewable.io:443/reviews/ratpack/ratpack/720#-Ju7CJ9itqcvp6cirKnY-r1-400)** ([raw file](https://github.com/ratpack/ratpack/blob/eed95a008fd7532c3d9f9a9b380ebc6c550634b3/ratpack-pac4j/src/main/java/ratpack/pac4j/RatpackPac4j.java#L400)):</sup>
This is a bit wack.

Why are we blocking here? Also, why are we masking errors from parse?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/ratpack/ratpack/720)

**Comment by johnrengelman** [comment]
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

<sup>**[ratpack-core/src/main/java/ratpack/parse/Parser.java, line 68 \[r1\]](https://reviewable.io:443/reviews/ratpack/ratpack/720#-Ju9bhQtZv0fwQbHV6_M)** ([raw file](https://github.com/ratpack/ratpack/blob/eed95a008fd7532c3d9f9a9b380ebc6c550634b3/ratpack-core/src/main/java/ratpack/parse/Parser.java#L68)):</sup>
The problem here is that the implementations call `Context.parse` which is not returning a promise...so how would resolve it to the actual type?

---

<sup>**[ratpack-pac4j/src/main/java/ratpack/pac4j/RatpackPac4j.java, line 400 \[r1\]](https://reviewable.io:443/reviews/ratpack/ratpack/720#-Ju7CJ9itqcvp6cirKnY-r1-400)** ([raw file](https://github.com/ratpack/ratpack/blob/eed95a008fd7532c3d9f9a9b380ebc6c550634b3/ratpack-pac4j/src/main/java/ratpack/pac4j/RatpackPac4j.java#L400)):</sup>
I'm not happy about this, but I did it because the Pac4J stuff is synchronous and I don't see a way to get around it. So I wrapped the implementation of the Pac4J stuff in a `blocking()` call to force this.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/ratpack/ratpack/720)

**Comment by alkemist** [comment]
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

<sup>**[ratpack-core/src/main/java/ratpack/parse/Parser.java, line 68 \[r1\]](https://reviewable.io:443/reviews/ratpack/ratpack/720#-Ju9bhQtZv0fwQbHV6_M)** ([raw file](https://github.com/ratpack/ratpack/blob/eed95a008fd7532c3d9f9a9b380ebc6c550634b3/ratpack-core/src/main/java/ratpack/parse/Parser.java#L68)):</sup>
Good point. This is awkward. Many implementations are not going to be async, but some are.  

One approach here would be to add a sync variant of parse() on context that takes a TypedData. The async one can just call this (i.e. map the request body promise).

---

<sup>**[ratpack-pac4j/src/main/java/ratpack/pac4j/RatpackPac4j.java, line 400 \[r1\]](https://reviewable.io:443/reviews/ratpack/ratpack/720#-Ju7CJ9itqcvp6cirKnY-r1-400)** ([raw file](https://github.com/ratpack/ratpack/blob/eed95a008fd7532c3d9f9a9b380ebc6c550634b3/ratpack-pac4j/src/main/java/ratpack/pac4j/RatpackPac4j.java#L400)):</sup>
You've got yourself in a tangle here. We are already returning the context async, there's no need to block.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/ratpack/ratpack/720)

**Comment by johnrengelman** [comment]
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

<sup>**[ratpack-pac4j/src/main/java/ratpack/pac4j/RatpackPac4j.java, line 400 \[r1\]](https://reviewable.io:443/reviews/ratpack/ratpack/720#-Ju7CJ9itqcvp6cirKnY-r1-400)** ([raw file](https://github.com/ratpack/ratpack/blob/eed95a008fd7532c3d9f9a9b380ebc6c550634b3/ratpack-pac4j/src/main/java/ratpack/pac4j/RatpackPac4j.java#L400)):</sup>
But we are updating the `Context.getBody()` method to return a `Promise` and the code inside `RatpackWebContext` needs the actual body. So how would we resolve that here since we can't change the the `WebContext` API to return `Promise`?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/ratpack/ratpack/720)

**Comment by johnrengelman** [comment]
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

<sup>**[ratpack-core/src/main/java/ratpack/http/internal/ContentTypeHandler.java, line 33 \[r1\]](https://reviewable.io:443/reviews/ratpack/ratpack/720#-Ju9crLZ4GQTVrRonnuT)** ([raw file](https://github.com/ratpack/ratpack/blob/eed95a008fd7532c3d9f9a9b380ebc6c550634b3/ratpack-core/src/main/java/ratpack/http/internal/ContentTypeHandler.java#L33)):</sup>
So, we should promote the `getContentType()` method from `TypedData` to `Request`?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/ratpack/ratpack/720)

**Comment by johnrengelman** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

<sup>**[ratpack-core/src/main/java/ratpack/parse/Parser.java, line 68 \[r1\]](https://reviewable.io:443/reviews/ratpack/ratpack/720#-Ju9bhQtZv0fwQbHV6_M)** ([raw file](https://github.com/ratpack/ratpack/blob/eed95a008fd7532c3d9f9a9b380ebc6c550634b3/ratpack-core/src/main/java/ratpack/parse/Parser.java#L68)):</sup>
Do we want to provide this overrided type signature on all the `parse` variants on `Context`?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/ratpack/ratpack/720)

**Comment by alkemist** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

<sup>**[ratpack-core/src/main/java/ratpack/http/internal/ContentTypeHandler.java, line 33 \[r1\]](https://reviewable.io:443/reviews/ratpack/ratpack/720#-Ju9crLZ4GQTVrRonnuT)** ([raw file](https://github.com/ratpack/ratpack/blob/eed95a008fd7532c3d9f9a9b380ebc6c550634b3/ratpack-core/src/main/java/ratpack/http/internal/ContentTypeHandler.java#L33)):</sup>
It should be on both.

---

<sup>**[ratpack-core/src/main/java/ratpack/parse/Parser.java, line 68 \[r1\]](https://reviewable.io:443/reviews/ratpack/ratpack/720#-Ju9bhQtZv0fwQbHV6_M)** ([raw file](https://github.com/ratpack/ratpack/blob/eed95a008fd7532c3d9f9a9b380ebc6c550634b3/ratpack-core/src/main/java/ratpack/parse/Parser.java#L68)):</sup>
Just this one:

https://github.com/ratpack/ratpack/blob/master/ratpack-core/src/main/java/ratpack/handling/Context.java#L656-656

---

<sup>**[ratpack-pac4j/src/main/java/ratpack/pac4j/RatpackPac4j.java, line 400 \[r1\]](https://reviewable.io:443/reviews/ratpack/ratpack/720#-Ju7CJ9itqcvp6cirKnY-r1-400)** ([raw file](https://github.com/ratpack/ratpack/blob/eed95a008fd7532c3d9f9a9b380ebc6c550634b3/ratpack-pac4j/src/main/java/ratpack/pac4j/RatpackPac4j.java#L400)):</sup>
Will push some code.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/ratpack/ratpack/720)

**Comment by alkemist** [comment]
Reviewed 8 of 31 files at r1, 6 of 8 files at r2, 20 of 22 files at r3.
Review status: 34 of 36 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

<sup>**[ratpack-core/src/main/java/ratpack/handling/Context.java, line 652 \[r3\]](https://reviewable.io:443/reviews/ratpack/ratpack/720#-JuEltglT6gA6_IAjbpy)** ([raw file](https://github.com/ratpack/ratpack/blob/54df79b9429204cce6449329165e161a019496be/ratpack-core/src/main/java/ratpack/handling/Context.java#L652)):</sup>
eh em.

---

<sup>**[ratpack-core/src/main/java/ratpack/http/internal/DefaultRequest.java, line 206 \[r3\]](https://reviewable.io:443/reviews/ratpack/ratpack/720#-JuEmc6WPuz7VxWqdOFb)** ([raw file](https://github.com/ratpack/ratpack/blob/54df79b9429204cce6449329165e161a019496be/ratpack-core/src/main/java/ratpack/http/internal/DefaultRequest.java#L206)):</sup>
Some duplication here re content type.

---

<sup>**[ratpack-core/src/main/java/ratpack/http/Request.java, line 128 \[r3\]](https://reviewable.io:443/reviews/ratpack/ratpack/720#-JuEmhfafuYZmPzRDALL)** ([raw file](https://github.com/ratpack/ratpack/blob/54df79b9429204cce6449329165e161a019496be/ratpack-core/src/main/java/ratpack/http/Request.java#L128)):</sup>
Should probably say what this is about, and specifically what happens if the request has no content-type header.

---

<sup>**[ratpack-core/src/test/groovy/ratpack/parse/ParserSpec.groovy, line 50 \[r3\]](https://reviewable.io:443/reviews/ratpack/ratpack/720#-JuEn0w0bfe59p2O3iwE)** ([raw file](https://github.com/ratpack/ratpack/blob/54df79b9429204cce6449329165e161a019496be/ratpack-core/src/test/groovy/ratpack/parse/ParserSpec.groovy#L50)):</sup>
Why is this var called type?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/ratpack/ratpack/720)

**Comment by alkemist** [comment]
Reviewed 8 of 31 files at r1, 6 of 8 files at r2, 22 of 22 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/ratpack/ratpack/720)

**Comment by johnrengelman** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

<sup>**[ratpack-core/src/main/java/ratpack/handling/Context.java, line 652 \[r3\]](https://reviewable.io:443/reviews/ratpack/ratpack/720#-JuEltglT6gA6_IAjbpy)** ([raw file](https://github.com/ratpack/ratpack/blob/54df79b9429204cce6449329165e161a019496be/ratpack-core/src/main/java/ratpack/handling/Context.java#L652)):</sup>
/me drops head.

---

<sup>**[ratpack-core/src/test/groovy/ratpack/parse/ParserSpec.groovy, line 50 \[r3\]](https://reviewable.io:443/reviews/ratpack/ratpack/720#-JuEn0w0bfe59p2O3iwE)** ([raw file](https://github.com/ratpack/ratpack/blob/54df79b9429204cce6449329165e161a019496be/ratpack-core/src/test/groovy/ratpack/parse/ParserSpec.groovy#L50)):</sup>
I think cause I misread it and though it was responding with the class type not the value.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/ratpack/ratpack/720)

**Comment by alkemist** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

<sup>**[ratpack-core/src/main/java/ratpack/handling/Context.java, line 652 \[r3\]](https://reviewable.io:443/reviews/ratpack/ratpack/720#-JuEltglT6gA6_IAjbpy)** ([raw file](https://github.com/ratpack/ratpack/blob/54df79b9429204cce6449329165e161a019496be/ratpack-core/src/main/java/ratpack/handling/Context.java#L652)):</sup>
The joys of code review.

---

<sup>**[ratpack-core/src/test/groovy/ratpack/parse/ParserSpec.groovy, line 50 \[r3\]](https://reviewable.io:443/reviews/ratpack/ratpack/720#-JuEn0w0bfe59p2O3iwE)** ([raw file](https://github.com/ratpack/ratpack/blob/54df79b9429204cce6449329165e161a019496be/ratpack-core/src/test/groovy/ratpack/parse/ParserSpec.groovy#L50)):</sup>
Should change it back.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/ratpack/ratpack/720)

**Comment by alkemist** [comment]
Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/ratpack/ratpack/720)

**Comment by alkemist** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

<sup>**[ratpack-core/src/test/groovy/ratpack/parse/ParserSpec.groovy, line 50 \[r3\]](https://reviewable.io:443/reviews/ratpack/ratpack/720#-JuEn0w0bfe59p2O3iwE)** ([raw file](https://github.com/ratpack/ratpack/blob/54df79b9429204cce6449329165e161a019496be/ratpack-core/src/test/groovy/ratpack/parse/ParserSpec.groovy#L50)):</sup>
Done

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/ratpack/ratpack/720)

**Comment by alkemist** [comment]
Reviewed 8 of 31 files at r1, 6 of 8 files at r2, 18 of 22 files at r3, 2 of 2 files at r4, 3 of 3 files at r5.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

<sup>**[ratpack-core/src/main/java/ratpack/handling/Context.java, line 652 \[r5\]](https://reviewable.io:443/reviews/ratpack/ratpack/720#-JuIzhqPm3IlB72M3bbi)** ([raw file](https://github.com/ratpack/ratpack/blob/50a544eb50d74bf58e5af7d77449d43c8f6540f4/ratpack-core/src/main/java/ratpack/handling/Context.java#L652)):</sup>
This really should explain when you'd use this method over the others. I'll fix it.

Good docs from here on out.

---

<sup>**[ratpack-core/src/main/java/ratpack/http/Request.java, line 130 \[r5\]](https://reviewable.io:443/reviews/ratpack/ratpack/720#-JuJ-0XMPfwAaFz5NeFe)** ([raw file](https://github.com/ratpack/ratpack/blob/50a544eb50d74bf58e5af7d77449d43c8f6540f4/ratpack-core/src/main/java/ratpack/http/Request.java#L130)):</sup>
I would have done this as:

{@link MediaType#isEmpty() an empty media type}

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/ratpack/ratpack/720)

**Pull request merged**